motion csv import

MaximilianKrambach opened the pull request:
A csv importer for motions . Still some things to do

(Part of #4074)

FinnStutzenstein commented on client/src/app/core/services/csv-export.service.ts:
```diff
@@ -179,4 +178,9 @@ export class CsvExportService {
         return this.translate.instant(temp);
     }
+
+    public exportDummy(data: string, filename: string): void {
```
Docs

FinnStutzenstein commented on client/src/app/site/motions/components/motion-import-list/motion-import-list.component.html:
```diff
@@ -0,0 +1,207 @@
+                <span translate>Motion block</span>
+            </div>
+        <ul>
+            <li translate>Identifier, reason, submitter, category, origin and motion block are optional and may be empty.
```
End tags missing.

FinnStutzenstein commented on client/src/app/site/motions/components/motion-import-list/motion-import-list.component.html:
```diff
@@ -0,0 +1,207 @@
+        <div>
+            <input type="file" accept="text" #fileInput (change)="onSelectFile($event)" />
+            &nbsp;
+            <span *ngIf="hasFile">{{ totalCount }}&nbsp;</span>
```
`<span *ngIf="hasFile">{{ totalCount }}&nbsp;<span translate>entries found</span>.</span>`

FinnStutzenstein commented on client/src/app/site/motions/components/motion-import-list/motion-import-list.component.html:
```diff
@@ -0,0 +1,207 @@
+            <mat-cell *matCellDef="let entry">
+                <div *ngIf="!entry.status_done">
+                    <div *ngIf="getInvalidReason(entry)">
+                        <mat-icon class="red-warning-text" matTooltip="{{getInvalidReason(entry)|translate}}">
```
Please reformat all interpolation tags and add spaces.

FinnStutzenstein commented on client/src/app/site/motions/components/motion-import-list/motion-import-list.component.ts:
```diff
@@ -0,0 +1,265 @@
+        this.setFilter();
+        this.importer.getNewEntries().subscribe(newEntries => {
+            me.dataSource.data = newEntries;
+            if (newEntries.length){
```
maybe a `me.hasFile = newEntries.length > 0;`.
Also check, if the `me` is necessary.

FinnStutzenstein commented on client/src/app/site/motions/components/motion-import-list/motion-import-list.component.ts:
```diff
@@ -0,0 +1,265 @@
+        super(titleService, translate, matSnackBar);
+        this.initTable();
+        this.selectedEncoding = this.encodings[0];
+        this.importer.errorEvent.subscribe( message => this.raiseError(message) );
```
` this.importer.errorEvent.subscribe(this.raiseError);` should work, becuase `raiseError` is a function that takes one argument.

FinnStutzenstein commented on client/src/app/site/motions/components/motion-import-list/motion-import-list.component.ts:
```diff
@@ -0,0 +1,265 @@
+    private setFilter(): void {
+        const me = this;
+        this.dataSource.filter = '';
+        this.dataSource.filterPredicate = function (
```
Please use an arrow function. This has the advantage, that you do not need the `me`, becuase an arrow function does not have it's own scope.

FinnStutzenstein commented on client/src/app/site/motions/components/motion-import-list/motion-import-list.component.ts:
```diff
@@ -0,0 +1,265 @@
+        if (!row.newMotion.title) {
+            return this.invalidReasons.title;
+        }
+        if (!row.newMotion.text) {
```
Make make `else if` out of this and if `this.invalidReasons` is not used anywhere else, move the strings into this function and remove `this.invalidReasons`.

FinnStutzenstein commented on client/src/app/site/motions/components/motion-import-list/motion-import-list.component.ts:
```diff
@@ -0,0 +1,265 @@
+        if (row.status_done) {
+            return 'import-done import-decided';
+        }
+        if (this.getInvalidReason(row)) {
```
Also some else if-structure would be nice.

FinnStutzenstein commented on client/src/app/site/motions/components/motion-import-list/motion-import-list.component.ts:
```diff
@@ -0,0 +1,265 @@
+     * @param inputString
+     */
+    private stripHtmlTags(inputString: string): string{
+        // TODO: allow some html elements like li p br
```
A plain preview is enough.

FinnStutzenstein commented on client/src/app/site/motions/components/motion-import-list/motion-import-list.component.ts:
```diff
@@ -0,0 +1,265 @@
+    public selectColSep(event: any): any {
+        this.importer.columnSeparator = event.target.value;
+    }
+
```
Remove empty line.

FinnStutzenstein commented on client/src/app/site/motions/models/view-motion.ts:
```diff
@@ -420,4 +433,5 @@ export class ViewMotion extends BaseViewModel {
             this._attachments
         );
     }
+
```
Please remove.

FinnStutzenstein commented on client/src/app/site/motions/models/view-motion.ts:
```diff
@@ -226,6 +226,19 @@ export class ViewMotion extends BaseViewModel {
         return this._attachments ? this._attachments : null;
     }
 
+    /**
```
For creation (what is done on import) please use `ViewCreateMotion`.

FinnStutzenstein commented on client/src/app/site/motions/models/view-motion.ts:
```diff
@@ -226,6 +226,19 @@ export class ViewMotion extends BaseViewModel {
+    /**
+     * Checks if the ViewMotion is valid: A motion needs a title and some content(text)
+     */
+    public get isValidMotion(): boolean {
```
`return this._motion && this.motion.title && this.motion.text;`

FinnStutzenstein commented on client/src/app/site/motions/motions-routing.module.ts:
```diff
@@ -20,9 +22,10 @@ const routes: Routes = [
     { path: ':id', component: MotionDetailComponent },
     { path: ':id/speakers', component: SpeakerListComponent },
-    { path: ':id/create-amendment', component: AmendmentCreateWizardComponent }
+    { path: ':id/create-amendment', component: AmendmentCreateWizardComponent },
```
Please undo

FinnStutzenstein commented on client/src/app/site/motions/services/category-repository.service.ts:
```diff
@@ -1,4 +1,5 @@
 import { Injectable } from '@angular/core';
+import { Observable, BehaviorSubject } from 'rxjs';
```
Pelase undo all changes here.

FinnStutzenstein commented on client/src/app/site/motions/services/motion-block-repository.service.ts:
```diff
@@ -112,4 +112,14 @@ export class MotionBlockRepositoryService extends BaseRepository<ViewMotionBlock
+     * @param title Strign to check for
+     */
+    public getMotionBlockByTitle(title: string): MotionBlock {
+        const filteredBlocks = this.DS.find(MotionBlock, block => block.title === title);
```
Just a return here should be enough. If the DS cannot find anything, this should be null.

FinnStutzenstein commented on client/src/app/site/motions/services/motion-csv-export.service.ts:
```diff
@@ -61,4 +61,17 @@ export class MotionCsvExportService {
     }
+
+    public exportDummyMotion(): void {
+        const dummyData = ['Identifier', 'Title', 'Text', 'Reason', 'Submitters',
```
Make from this first map statement a variable called `headerrow`.
then make a list like this: `rows = [headerrow, 'row 1', 'row 2', ...].
Return a `rows.join('\n')`

FinnStutzenstein commented on client/src/app/site/motions/services/motion-import.service.ts:
```diff
@@ -0,0 +1,435 @@
+
+export interface NewMotionEntry {
+    newMotion: CreateMotion;
+    newViewMotion: ViewMotion;
```
Use `ViewCreateMotion`.

FinnStutzenstein commented on client/src/app/site/motions/services/motion-import.service.ts:
```diff
@@ -0,0 +1,435 @@
+@Injectable({
+    providedIn: 'root'
+})
+export class MotionImportService {
```
Some docs would be nice ;)

FinnStutzenstein commented on client/src/app/site/motions/services/motion-import.service.ts:
```diff
@@ -0,0 +1,435 @@
+    ) {
+        this.reader = new FileReader();
+        const me = this;
+        this.reader.onload = function(event: any): void {
```
Use an arrow function and remove the `me`.

FinnStutzenstein commented on client/src/app/site/motions/services/motion-import.service.ts:
```diff
@@ -0,0 +1,435 @@
+                    )
+                );
+            }
+            await Promise.all(promises);
```
This should be outside of the loop.

FinnStutzenstein commented on client/src/app/site/motions/services/motion-import.service.ts:
```diff
@@ -0,0 +1,435 @@
+        if (this.columnSeparator) {
+            papaConfig.delimiter = this.columnSeparator;
+        }
+        const entry_lines = this.papa.parse(file, papaConfig).data;
```
camelCase

FinnStutzenstein commented on client/src/app/site/motions/services/motion-import.service.ts:
```diff
@@ -0,0 +1,435 @@
+            papaConfig.delimiter = this.columnSeparator;
+        }
+        const entry_lines = this.papa.parse(file, papaConfig).data;
+        if ( !this.checkHeader(entry_lines.shift()) ) {
```
As discussed: just throw an error, if the min-length (3, identifier, title, text is always required) is not statisfied. Throw a better error message.

MaximilianKrambach replied:
done by checkHeader: snackbar if just a warning, error else

FinnStutzenstein commented on client/src/app/site/motions/services/motion-import.service.ts:
```diff
@@ -0,0 +1,435 @@
+                let category: { category_id?: number, category_name?: string, newCategory?: string };
+                let motion_block: { block_id?: number, block_name?: string, newBlock?: string };
+
+                for (let idx = 0; idx < this.expectedHeader.length; idx++ ) {
```
Note: You need to have the limit of the loop to be the minimum of expectedHeaders and "real" headers.

FinnStutzenstein commented on client/src/app/site/motions/services/motion-import.service.ts:
```diff
@@ -0,0 +1,435 @@
+        };
+        this._entries.forEach(entry => {
+            summary.total += 1;
+            if (entry.status_done) {
```
use else ifs.

FinnStutzenstein commented on client/src/app/site/motions/services/motion-repository.service.ts:
```diff
@@ -583,4 +583,22 @@ export class MotionRepositoryService extends BaseRepository<ViewMotion, Motion>
+     * in the data
+     * @param sharedValues properties that must be equal to consider it a duplicate
+     */
+    public getMotionDuplicates(viewMotion: ViewMotion, sharedValues: (keyof { [x: string]: Motion})[]): ViewMotion[] {
```
`(keyof Motion)[]`

FinnStutzenstein commented on client/src/app/site/users/services/group-repository.service.ts:
```diff
@@ -1,4 +1,5 @@
 import { Injectable } from '@angular/core';
+import { BehaviorSubject, Observable } from 'rxjs';
```
Undo all changes here.

tsiegleauq commented on client/src/app/site/motions/components/motion-list/motion-list.component.html:
```diff
@@ -152,6 +152,10 @@
         </button>
+        <button mat-menu-item *osPerms="'motions.can_manage'" routerLink="import">
+            <mat-icon>save_alt</mat-icon>
+            <span translate>Import ...</span>
```
"..." does not need to be translated. Only translate "Import"

tsiegleauq commented on client/src/app/site/motions/models/csv-create-motion.ts:
```diff
@@ -0,0 +1,154 @@
+ */
+export class CsvCreateMotion extends CreateMotion {
+
+    public csvCategory: CsvMapping;
```
documentation (whole class) + prettify required

tsiegleauq commented on client/src/app/site/motions/models/csv-create-motion.ts:
```diff
@@ -0,0 +1,154 @@
+    /**
+     * takes a list of motion block maps to update the current csvMotionblock.
+     * Returns the amount of entries that remain unmatched
+     * @param motionBlocks
```
line spacing

tsiegleauq commented on client/src/app/site/motions/models/csv-create-motion.ts:
```diff
@@ -0,0 +1,154 @@
+    /**
+     * takes a list of category maps to update the current csv_category.
+     * Returns the amount of entries that remain unmatched
+     * @param solveCategory
```
line spacing

tsiegleauq commented on client/src/app/site/motions/services/motion-import.service.ts:
```diff
@@ -0,0 +1,418 @@
+import { MatSnackBar } from '@angular/material';
+import { CsvCreateMotion, CsvMapping } from '../models/csv-create-motion';
+
+export interface NewMotionEntry {
```
doc missing

tsiegleauq commented on client/src/app/site/motions/services/motion-import.service.ts:
```diff
@@ -0,0 +1,418 @@
+     * @param translate Translation service
+     * @param papa External csv parser (ngx-papaparser)
+     */
+    public constructor( private repo: MotionRepositoryService,
```
file needs to be prettified.
matSnackbar is missing in constructor doc:

tsiegleauq commented on client/src/app/site/users/services/user-import.service.ts.alt:
```diff
@@ -0,0 +1,235 @@
+@Injectable({
+    providedIn: 'root'
+})
+export class UserImportService {
```
should this file really be there?
`ts.alt` ;-)

tsiegleauq commented on client/src/app/site/users/services/user-repository.service.ts:
```diff
@@ -195,4 +198,48 @@ export class UserRepositoryService extends BaseRepository<ViewUser, User> {
+    public findUserByName(name: string) : ViewUser[] {
+        const results : ViewUser[] = [];
+        const users = this.DS.getAll(User).filter(user => {
+            if (user.full_name === name || user.short_name === name){
```
run prettier on file

tsiegleauq commented on client/src/app/site/users/users-routing.module.ts:
```diff
@@ -4,6 +4,8 @@ import { UserListComponent } from './components/user-list/user-list.component';
 import { UserDetailComponent } from './components/user-detail/user-detail.component';
 import { GroupListComponent } from './components/group-list/group-list.component';
 import { PasswordComponent } from './components/password/password.component';
+//  import { UserImportListComponent } from './components/user-import/user-import-list.component';
```
remove

tsiegleauq commented on client/src/app/site/users/users-routing.module.ts:
```diff
@@ -32,8 +34,7 @@ const routes: Routes = [
         // data: { basePerm: 'users.can_manage' }
-    },
-    {
+    }, {
```
was prettier doing this? Looks a little odd

MaximilianKrambach commented on client/src/app/site/motions/components/motion-import-list/motion-import-list.component.html:
```diff
@@ -0,0 +1,232 @@
+            </mat-form-field>
+        </div>
+        <div>
+            <input type="file" accept="text" #fileInput (change)="onSelectFile($event)" />
```
the drag and drop of mediafiles didn't work here, so I used the first javascript thingy available (which is input type file). I changed it to a fake mat-button as you suggested.

And I really don't see the advantages of such a dragdrop

tsiegleauq commented on client/src/app/site/motions/components/motion-import-list/motion-import-list.component.spec.ts:
```diff
@@ -0,0 +1,25 @@
+
+  beforeEach(async(() => {
+    TestBed.configureTestingModule({
+      declarations: [ MotionImportListComponent ]
```
To get the tests running:
imports: [E2EImportsModule]

MaximilianKrambach commented:
It is now in a (rebased on master) updated state, where the secondary imports should work, too.

I removed the CsvCreateMotion model, and moved the logic from there to either the CsvMotionImportService or the (newly created) ViewCsvCreateMotion. I hope that's an improvement

MaximilianKrambach commented:
fixed submitters, tested motionblocks too, attempted improving 'new item' design (still not happy with that)

tsiegleauq commented on client/src/app/site/motions/components/motion-import-list/motion-import-list.component.ts:
```diff
@@ -0,0 +1,274 @@
+        return this.importer && this.hasFile ? this.importer.summary.total : null;
+    }
+
+    public get encodings(): ValueLabelCombination[] {
```
documentation missing

tsiegleauq commented on client/src/app/site/motions/components/motion-import-list/motion-import-list.component.ts:
```diff
@@ -0,0 +1,274 @@
+        return this.importer.encodings;
+    }
+
+    public get columnSeparators(): ValueLabelCombination[] {
```
documentation missing

tsiegleauq commented on client/src/app/site/motions/components/motion-import-list/motion-import-list.component.ts:
```diff
@@ -0,0 +1,274 @@
+
+    /**
+     * Returns the appropiate css class for a row according to the import state
+     * TODO
```
If this is an open TODO, describe what is left to be done

tsiegleauq commented on client/src/app/site/motions/components/motion-import-list/motion-import-list.component.html:
```diff
@@ -47,6 +47,14 @@
             </mat-select>
         </mat-form-field>
+        <mat-form-field>
+            <mat-label translate> Text separator</mat-label>
```
one whitespace too much.
(this might translate ` Text seperator` with additional whitespace at the beginning

tsiegleauq commented on client/src/app/site/motions/components/motion-import-list/motion-import-list.component.html:
```diff
@@ -192,7 +200,8 @@
                         <span *ngFor="let submitter of entry.newMotion.csvSubmitters">
                             {{ submitter.name
-                            }}<mat-icon class="newBadge" color="accent" inline *ngIf="!submitter.id">add</mat-icon>&nbsp;
+                            }}<mat-icon class="newBadge" color="accent" inline *ngIf="!submitter.id">add</mat-icon
```
strange indentation.

tsiegleauq commented on client/src/app/site/motions/components/motion-import-list/motion-import-list.component.html:
```diff
@@ -211,7 +220,9 @@
                         {{ entry.newMotion.csvCategory.name }}
-                        <mat-icon class="newBadge" color="accent" inline *ngIf="!entry.newMotion.csvCategory.id">add</mat-icon>&nbsp;
+                        <mat-icon class="newBadge" color="accent" inline *ngIf="!entry.newMotion.csvCategory.id"
+                            >add</mat-icon
```
indentation looks odd. Try adding a new line before "add"

tsiegleauq commented on client/src/app/site/motions/components/motion-import-list/motion-import-list.component.html:
```diff
@@ -235,7 +246,9 @@
                         {{ entry.newMotion.csvMotionblock.name }}
-                        <mat-icon class="newBadge" color="accent" inline *ngIf="!entry.newMotion.csvMotionblock.id">add</mat-icon>
+                        <mat-icon class="newBadge" color="accent" inline *ngIf="!entry.newMotion.csvMotionblock.id"
+                            >add</mat-icon
```
Looks odd, try add new line before odd

MaximilianKrambach commented:
@tsiegleauq The issue (not clearing newUsers etc. on new file parsing) is fixed now, the documentation issues addressed, rebased a some older commits of this PR; rebased on master

MaximilianKrambach commented:
@tsiegleauq I rebased it and had the new prettier running; everything passing now
